`storage`: add transactional batch utilities to storage layer & slightly rework compaction logic [SC&R #4]

WillemKauf opened the pull request:
This PR is the 4th in the series of safe compaction & recovery [SC&R] PRs.

This PR does two notable things:

1. Adds `has_transaction_batches` to the `index_state`. This is a definitive flag for a `segment` containing any transactional data batches (i.e control batches or raft data batches with a transactional bit set). This will be used in the future to make scheduling decisions with regards to compaction of a `segment`, similar to how it is currently performed with tombstones (i.e if a `segment` contains tombstones, we know about it via `index_state::may_have_tombstone_records` and we force self compaction if these become eligible for removal).
2. Reworks the definition of `is_compactible()` with a slightly new definition, `is_filterable()`. The key point here is that in the future, transactional control batches will be eligible for cleanup once their deletion window passes (`self_compact_timestamp + delete.retention.ms`). Previously, these batches were not considered "compactible"- meaning, we wouldn't even try to filter them through the `copy_data_segment_reducer`. Though these batches are still not "compactible" (i.e eligible for removal due to the existence of a record with the same key later in the log), they are removable by this deletion window- meaning they need to be considered as "filterable". To this effect, the logic of `should_keep()` and `copy_data_segment_reducer::filter()` is altered slightly to allow these batches into the code path that could potentially filter them out. "compactible" batches are still not considered by `should_keep::is_latest_key()`- they are either removed by deletion logic (this is still currently left unimplemented for transactional batches in this PR), or kept.


## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

WillemKauf committed:
`storage`: add `has_transaction_batches` to `index_state`

Answers the question of "does this segment have any transactional batches
in it?"

Just like `may_have_tombstone_records`, we need this in order to determine
whether or not a `segment` should undergo compaction in a timely manner
in order to remove transaction batches within. Just like the presence of
removable tombstones, the presence of removable transactional batches will act
as a forcing function for self compaction to occur (even if the `segment` is
already self compacted).

This is a definite flag, because we can quickly query whether a batch
contains transactional data or not in the append path (i.e we can default
this to `false` and only set it to `true` if a transactional batch is appended
to the `segment`).

However, this means that existing `segment`s on disk have a default value of
`false` for this flag (and may not guarantee correctness with
`has_transaction_batches`).

WillemKauf committed:
`storage`: split compaction logic into `is_compactible()` & `is_filterable()`

The major motivation for this change is the fact that transactional batches
will be considered "filterable" in the future, despite not being "compactible".

After the changes in this PR, the property "compactible" refers to the fact
that a record may be removed due to a record with the same key existing later
in the log (de-duplication). "Compactible" records should be indexed in the
`.compaction_index` file for a `segment`, and they should be considered in
`should_keep::is_latest_key()`.

On the other hand, the property "filterable" means that a record may be removed
due to either de-deduplication OR by other means during the compaction process.
Most notably, the removal of transaction batches past the horizon set by
`delete.retention.ms`. Filterable batches should be processed by `should_keep()`,
but some filterable batches should not be considered in
`should_keep::is_latest_key()` (aforementioned transaction batches which
should only be removed by `delete.retention.ms`, and _not_ de-duplication).

The property `num_compactible_records_appended` in the `index_state` now refers
to filterable records, not just compactible records. However, the naming
throughout the `storage` layer for this function and its related call sites
are left the same to prevent this slight implementation detail/nuance from
leaking out too far outside of compaction.

WillemKauf commented on src/v/storage/segment_utils.h:
```diff
@@ -388,8 +397,10 @@ ss::future<bool> should_keep(
     auto& header = b.header();
-    if (header.attrs.is_control()) {
-        has_tx_batches = true;
+    if (!is_compactible(ntp, header)) {
```
Unfortunately, we are not getting around to removing the `ntp` from this path yet due to the condition here:

https://github.com/redpanda-data/redpanda/blob/8b6866b266962cd8668a7553591a430401d91b70/src/v/storage/segment_utils.h#L266-L268

These `tx_fence` batches are still not considered "compactible", they are considered "filterable" with these changes. @bharathv, do you think considering them "compactible" for all topics would be fine in the future (due to presence of transactional bit in raft data)? That would allow us to remove the `ntp` from this path.

bharathv replied:
hmm I commented above.. IIUC with these changes a transactional CO batch (eg: `group_fence_tx`) results in is_compactible() = true (is_control() == true, is_compactible_control_batch()== true && is_filterable() == true), so it will be indexed in the offset map like any other batch and results in non determinism of removal?

WillemKauf replied:
The transactional CO batches haven't had a change in behavior as far as I understand it with the split of `compactible` and `filterable` (they were still considered compactible before and would be indexed in the offset map). I'm referring to non-CO `tx_fence` batches here (and bringing up the point that it would be nice to remove the `ntp` from this path if all `tx_fence` batches in internal and user topics were considered "compactible"

WillemKauf commented on src/v/storage/segment_utils.cc:
```diff
@@ -800,6 +806,33 @@ ss::future<compaction_result> self_compact_segment(
           gclog.debug,
           "detected {} is already compacted",
           s->path().to_compacted_index());
+        if (!s->has_self_compact_timestamp()) {
```
This check here (and the entire codeblock) is only valid if `self_compact_timestamp` and `has_transactional_batches` makes it into the same initial release.

bharathv replied:
This block seems a bit risky? (lot of unexpected IO from rescan of compacted segments, post upgrade)

Can't we simply exclude such segments from the safe recovery logic?
Since these segments were produced and compacted before this new feature was introduced, no guarantee for safe recovery which is the status quo.

Effectively, we should only consider segments that have self_compaction_timestamp set for safe recovery. This implies that has_transaction_batches should be 100% reliable when self_compaction_timestamp is present, so both fields should be introduced together. Which requires this patch to be released into 25.2.

(did I miss something :D)

WillemKauf replied:
>This block seems a bit risky? (lot of unexpected IO from rescan of compacted segments, post upgrade)

Yes, seems like a lot of potential re-reads, though, we don't have to decompress or iterate over records in this path so I hoped it would be a bit less straining on the system... but at the end of the day, it is re-reads over the existing `segment` files.

> Can't we simply exclude such segments from the safe recovery logic Since these segments were produced and compacted before this new feature was introduced, no guarantee for safe recovery which is the status quo.

I would be open to this. (I.e removing this block of code)

> Effectively, we should only consider segments that have self_compaction_timestamp set for safe recovery. 

Agreed

> This implies that has_transaction_batches should be 100% reliable when self_compaction_timestamp is present, so both fields should be introduced together. Which requires this patch to be released into 25.2.

Yes, `has_transaction_batches` must be reliable, and yes, this patch must be released alongside `self_compact_timestamp`.

WillemKauf replied:
Removed this block of code.

bharathv commented on src/v/storage/segment.cc:
```diff
@@ -631,22 +631,33 @@ ss::future<append_result> segment::do_append(const model::record_batch& b) {
+              "Marking index for segment {} as has transaction batches",
+              filename());
+        }
+        _idx.set_has_transaction_batches(true);
```
nit: can we move this in to the if above? (set only if unset) (one less call)

WillemKauf replied:
Moved.

bharathv commented on src/v/storage/segment_utils.h:
```diff
@@ -373,6 +374,10 @@ ss::future<bool> should_keep(
             may_have_tombstone_records = true;
         }
 
+        if (b.contains_transactional_data()) {
```
nit: should we remove this branch altogether? compaction_placeholder_enabled = true always and the entire block is deadcode (the nesting is confusing)

WillemKauf replied:
>compaction_placeholder_enabled = true always

What about clusters in which the feature is disabled?

bharathv replied:
how? This was introduced in 24.1.0 and given the linear upgrade path requirements, a cluster can only be upgraded to this code from 25.1.0 and that makes the feature enabled always?

WillemKauf replied:
We allow users to disable features through the admin API, no? In normal operations I agree the check would likely be unnecessary but it seems necessary to keep the checks around.

bharathv replied:
true, but this is not a user facing feature, unlikely anyone even touches it.. pretty sure we did something like for other features to GC dead code.. either way probably ok to do it in a future patch.

bharathv commented on src/v/storage/segment_utils.h:
```diff
@@ -382,12 +387,22 @@ ss::future<bool> should_keep(
     }
 
+    if (b.contains_transactional_data() && keep) {
+        has_tx_batches = true;
```
nit: has_tx_batches = has_tx_batches || (b.contains_transactional_data() && keep);

WillemKauf replied:
How about one better with `has_tx_batches |= (b.contains_transactional_data() && keep);`

bharathv commented on src/v/storage/segment_utils.h:
```diff
@@ -272,23 +266,37 @@ inline bool is_compactible_control_batch(
+// batches, as well as all batches that shift offset translation. These batches
+// shouldn't even be considered in `should_keep()`, they should be kept
+// unconditionally.
+inline bool is_filterable(const model::record_batch_type& t) {
```
nit: pass batch_type by copy.

bharathv commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -148,9 +148,8 @@ model::record_batch copy_data_segment_reducer::make_placeholder_batch(
-    // offset translation
-    if (!is_compactible(_ntp, batch)) {
+    // do not filter non-removable batch types under any circumstances
+    if (!is_filterable(batch.header().type)) {
```
with this change, I'm wondering if there is a probabilistic angle to removing control batches. Clearly they are not indexed in the offset map because they are not compactible, but maybe_keep_offset() can still have false positives? (due to usage of using roaring bitmap for offsets)

Particularly I'm thinking about consumer offsets transactional batches (which must be removed deterministically for correctness).

WillemKauf replied:
`is_compactible(_ntp, batch)` for any of `group_fence_tx`, `group_prepare_tx`, `group_abort_tx` and `group_commit_tx` before this change `== true`.

After this change, `is_filterable(batch.header().type)` for any of `group_fence_tx`, `group_prepare_tx`, `group_abort_tx` and `group_commit_tx` is `== true`, and `is_compactible(_ntp, batch)` is still `== true`.

What change in behavior here is concerning?

bharathv replied:
oops, mybad I got confused.. I misinterpreted existing code (which is already probabilistic in nature to some extent, but that can be a separate PR).

WillemKauf commented:
Force push to:
* Move call to `_idx.set_has_transaction_batches` in `segment.cc`
* Remove `contains_tx_consumer` block in `segment_utils.cc`
* Change signature to `inline bool is_filterable(model::record_batch_type t)` (pass by value)
* Change assignment logic of `may_have_tombstone_records` and `has_tx_batches` in `should_keep()`

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--20aad970-057a-4cf9-ac59-ebb7dec8a9ff-->
<!--TEST RESULTS-->
<!--d9aba917-41b0-4b25-af7a-d40bf8761065-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68484">build#68484</a></summary>


| test_class               | test_method                        | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-----------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_migrated_topic_data_integrity | {"params": {"cancellation": {"dir": "in", "stage": "executing"}, "use_alias": true}, "transfer_leadership": true}                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68484#0197d686-f9af-4e94-82d4-3415c535c017 | FLAKY         | 20/21    | upstream reliability is '99.71671388101983'. current run reliability is '95.23809523809523'. drift is 4.47862 and the allowed drift is set to 50. The test should PASS |
| LeadershipPinningTest    | test_leadership_pinning_sanctions  |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68484#0197d686-f9af-4a20-90ad-96ae6ca75c8c | FLAKY         | 20/21    | upstream reliability is '99.71910112359551'. current run reliability is '95.23809523809523'. drift is 4.48101 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations               | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68484#0197d682-58a9-4590-a12e-afe17a6e5721 | FLAKY         | 19/21    | upstream reliability is '99.56709956709958'. current run reliability is '90.47619047619048'. drift is 9.09091 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68535">build#68535</a></summary>


| test_class                  | test_method                              | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:-----------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68535#0197e6a6-5ace-4dbc-95bb-e297dfa94034 | FLAKY         | 19/21    | upstream reliability is '98.75311720698254'. current run reliability is '90.47619047619048'. drift is 8.27693 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68535#0197e6a6-5ace-4dbc-95bb-e297dfa94034 | FLAKY         | 19/21    | upstream reliability is '99.2156862745098'. current run reliability is '90.47619047619048'. drift is 8.7395 and the allowed drift is set to 50. The test should PASS   |
</details>
<!--/body-->

bharathv commented on src/v/storage/segment_utils.h:
```diff
@@ -382,11 +387,16 @@ ss::future<bool> should_keep(
-    if (r.is_tombstone() && keep) {
-        may_have_tombstone_records = true;
-    }
+    may_have_tombstone_records |= (r.is_tombstone() && keep);
```
this is bitwise OR right, it doesn't shortcircuit like logical OR I think.. the result is probably correct because the operands are implicitly converted to 0/1.

WillemKauf replied:
Hm yeah. Probably _slightly_ more efficient to keep the version with short circuiting. The result is correct either way though.

bharathv reviewed (commented):
lgtm, main question.. should we remove compaction_placeholder_feature pluming in this patch to make things simple.. otherwise good to go.

WillemKauf committed:
`storage`: add `can_discard()` function

Certain records/batches will always be considered removable during
compaction.

These include `compactible_control_batches()` such as those found in the
`__consumer_offsets` topic, and tombstone records past the delete time horizon
set by `delete.retention.ms`.

In the future, additional logic handling transactional batches shall also be
added to the `can_discard()` function.
